crates/loam-cli/src/commands/dev/mod.rs

```diff
@@ -0,0 +1,219 @@
+
+enum Message {
+    FileChanged,
+}
+
+#[derive(Parser, Debug, Clone)]
+#[group(skip)]
+pub struct Cmd {
+    #[command(flatten)]
+    pub build_cmd: build::Cmd,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Watcher(#[from] notify::Error),
+    #[error(transparent)]
+    Build(#[from] build::Error),
+    #[error("IO error: {0}")]
+    Io(#[from] std::io::Error),
+}
+
+fn canonicalize_path(path: &Path) -> PathBuf {
+    if path.as_os_str().is_empty() {
+        env::current_dir().unwrap_or_else(|_| PathBuf::from("."))
+    } else if path.components().count() == 1 {
+        // Path is a single component, assuming it's a filename
+        env::current_dir()
+            .unwrap_or_else(|_| PathBuf::from("."))
+            .join(path)
+    } else {
+        fs::canonicalize(path).unwrap_or_else(|_| path.to_path_buf())
+    }
+}
+
+#[derive(Clone)]
+struct Watcher {
+    root_env: Arc<PathBuf>,
+    packages: Arc<Vec<PathBuf>>,
+}
+
+impl Watcher {
+    pub fn new(root_env: &Path, packages: &[PathBuf]) -> Self {
+        Self {
+            root_env: Arc::new(canonicalize_path(root_env)),
+            packages: Arc::new(packages.iter().map(canonicalize_path).collect()),
+        }
+    }
+
+    pub fn is_watched(&self, path: &Path) -> bool {
+        let path = canonicalize_path(path);
+        self.packages.iter().any(|p| path.starts_with(p))
+    }
+
+    pub fn is_env_toml(&self, path: &Path) -> bool {
+        canonicalize_path(path) == *self.root_env
+    }
+}
+
+fn is_temporary_file(path: &Path) -> bool {
+    const IGNORED_EXTENSIONS: &[&str] = &["tmp", "swp", "swo"];
+    let file_name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
+
+    // Vim and vscode temporary files
+    if path
+        .extension()
+        .and_then(|ext| ext.to_str())
+        .map_or(false, |ext| {
+            IGNORED_EXTENSIONS
+                .iter()
+                .any(|&ignored| ext.eq_ignore_ascii_case(ignored))
+        })
+    {
+        return true;
+    }
+
+    // Vim temporary files
+    if file_name.ends_with('~') {
+        return true;
+    }
+
+    // Emacs temporary files
+    if file_name.starts_with('#') && file_name.ends_with('#') {
+        return true;
+    }
+
+    // Add more patterns for other editors as needed
+
+    false
+}
+
+impl Cmd {
+    pub async fn run(&mut self) -> Result<(), Error> {
+        let (tx, mut rx) = mpsc::channel::<Message>(100);
+        let rebuild_state = Arc::new(Mutex::new(false));
+        let workspace_root: &Path = self
+            .build_cmd
+            .manifest_path
+            .parent()
+            .unwrap_or_else(|| Path::new("."));
+        let env_toml_path = workspace_root.join("environments.toml");
+
+        let packages = self
+            .build_cmd
+            .list_packages()?
+            .into_iter()
+            .map(|package| PathBuf::from(package.manifest_path.parent().unwrap().as_str()))
```

Comment:
Isn't `parent` an `Option<&Path>`?
```suggestion
            .map(|package| package.manifest_path.parent().unwrap().into_path_buf())
```

Reply:
Seems the `to_path_buf` method returns a `Utf8PathBuf` which we'd then have to refactor alot of the code to pass that type. Is that what you'd prefer?

Reply:
This is fine for now. Again, we can return with `camino` later to clean up all path stuff.